text2troff_table.1
./" Manpage for your_function
.TH TEXT2TROFF_TABLE 1 "18 December 2023" "1.0"
.
.
.
.SH NAME
text2troff_table \(en converts tabulated lines in a plain-text-file into TROFF formatted tables by inserting table requests.
.
.
.
.SH SYNOPSIS
text2troff_table [\fIoptions\fP] [\fIinput-file\fP]
.
.
.
.SH DESCRIPTION
text2troff_table is a bash script that uses \f2awk\fP filtering to convert tabulated lines
in a plain-text-file into TROFF formatted tables by inserting table requests and ms macros.
All lines not interpreted as table lines are passed on unchanged.
The result is sent to standard output,
and can be processed further with available `troff' tools with `ms' macro package, preferrably GNU groff_ms(7),
to produce typeset PostScript-, PDF-, HTML- or terminal (`nroff') output.
As an alternative to processing a text-file, text2troff_table can also read (text) input from a pipe.
.
.
.
.SH INPUT TEXT REQUIREMENTS
In order to be interpreted as a table, lines must appear clustered as a continuous series of two (2)
or more text lines, each containing at least two (2) non-adjacent \f2tab\fP characters,
excluding any leading or trailing \f2tab\fPs per line.
This means that a minimum of three (3) text columns is required to be interpreted as a table.
.LP
Any continuous series of one or more \f2tab\fPs within a text line is treated as \f3one\fP
delimiter between adjacent table-`cells' c.q. columns.
This enables the use of multiple \f2tab\fPs in the input text, for instance to allign text
belonging to the same column for the sake of readibility.
As a consequence,
table cells supposed to be `empty' shall be filled by at least a \f2space\fP or e.g. a \f2dash\fP (\fC-\fP)
or any other chosen character.
Another consequence is that a \f2tab\fP character itself can \(en by definition \(en
never be part of the contents of a table cell.
.br
All text intended to be in a single table row shall be placed on the same text line.
.SH OPTIONS
text2troff_table takes the following options:
.IP "\f3-h\fP"
Print `usage' text for help.
.IP "\f3-e\fP"
Expand tables to full width of the running text margins.
.br
Without this option, tables are centered with respect to the running text margins and not wider than
strictly required by content.
.IP "\f3-s\fP"
Force (`squeeze') tables containing too many columns and/or unbreakable text to fit within running
text margins, by adding invisible break characters \: to the contained text.
.
.
.SH "TEXT TRANSFORMATION TO TROFF FORMAT"
First, text2troff_table removes any amount of \f2tab\fPs at the beginning or end of each line,
and it replaces any contiguous series of \f2tab\fPs within a text line by just one \f2tab\fP.
Not affected though are text lines within `displays',
as marked by presence of ms macros \f3\f8.DS\f1 (`display starts') and \f3\f8.DE\f1
(`display ends').
Also all ISO-8859 characters are converted to UTF-8,
and \f2return\fP characters (\fC\\r\fP) are removed.
.LP
Next, text2troff_table treats any contiguous cluster of two (2) or more text lines as a table,
if each contains at least two (2) non-adjacent \f2tab\fP characters.
The table is terminated as soon as a line passes that doesn't meet above conditions.
At any following cluster of lines that again complies, a next table starts.
.LP
Per table, the column quantity is the maximum number of tab-delimited cells
found in any of its lines.
.LP
The table-number is not hard-coded but instead driven by a TROFF register variable,
which is automatically incremented during interpretation by groff_ms.
.LP
Default font size within the tables is set at the default point size minus 20%.
Latter value is defined by a second TROFF register variable.
To set both variables,
text2troff_table inserts the following two lines at the top of the input text:\f1
.LP
.TS
lw(30) lw(40).
T{
.nf
\&\fC.nr table_nr 0
\&.nr ps_decr \\n[.s]*2/10\fP
.fi
T}	T{
.nf
Initialize new number register `table_nr' at 0
Set new number register `ps_decr' at 20% of point size
.fi
T}
.TE
.sp 0.6
At the top of each table, two lines with ms macro \f3\f8.LP\f1 (`paragraph starts') and
\f3\f8.TS H\f1 (`table with header starts') respectively are inserted,
as well a number of requests defining the table header and other table properties.
.sp
Per table line, each individual table cell as delimited by adjacent tab(s) is:
.IP -
prefixed by \f3\f8T{\f1 and a \f2newline\fP character, and
.IP -
postfixed by a \f2newline\fP character and \f3\f8T}\f1
.LP
This formatting allows the text to be wrapped within the cell by breaking at \f2space\fPs and syllable
ends when processed by groff_ms.
.LP
Concluding each table, two lines with ms macro \f3\f8.TE\f1 (`table ends') and \f3\f8.LP\f1
(`paragraph starts') respectively are inserted.
.sp 0.5
For example, the original text:
.LP
.IP
.nf
\fCAfter this line of main text, next lines represent a table:
first cell	\f2 tab \fP	second cell: long text 	\f2 tab \fP  third cell
fourth cell	\f2 tab \fP	fifth cell    \f2 tab \fP    	\f2 tab \fP  sixth cell
Here the main text continues after the table ...\f1
.fi
.sp 0.5
.LP
is converted by text2troff_table the following way:
.LP
.IP
.nf
\fCAfter this line of main text, next lines represent a table:
\&.LP                  	\f1Macro `paragraph starts'\fP
\&.nr table_nr +1      	\f1Number register `table_nr' is incremented by 1\fP
\&.ps -\\n[ps_decr]    	\f1Point size minus decrement value being 20% of point size\fP
\&.TS H                	\f1Request `table with header starts'\fP
\&Tabel \\n[table_nr]  	\f1Header text and incremented number register `table_nr'\f2
\&table_type\fC, allbox;   	\f1Requests `\f2table_type\fP' and `box around every table cell'\fC
\&c s s                	\f1Header cell centered (c), spans (s) each other column \fP
\&l l l .              	\f1Each cell of next rows left (l) alligned \fP
\&T{
\&first cell           	\f[1]First row of table cells\fC
\&T}\f2 tab \fPT{
\&second cell: long text
\&T}\f2 tab \fPT{
\&third cell
\&T}
\&T{
\&fourth cell          	\f1Second row of table cells\fP
\&T}\f2 tab \fPT{
\&fifth cell
\&T}\f2 tab \fPT{
\&sixth cell
\&T}
\&.TE                  	\f1Request `table ends'\fP
\&.LP                  	\f1ms macro `paragraph starts'\fP
\fCHere the main text continues after the table ...\fP
.fi
.sp 0.5
.LP
The \f2table_type\fP can be:
.IP "\fCcenter\fP"
In this default case, groff_ms will center the table between the text margins, at a width no more
than strictly required by table contents;
.br
.IP "\fCexpand\fP"
if option \f3-e\fP is given, or if option \f3-s\fP is given and table exceeds text margins,
groff_ms will expand the table to full width between the text margins.
.LP
if option \f3-s\fP is given, invisible break characters \\: are inserted in the table cell
text after every \f2n\fP characters \(en with \f2n\fP depending on on table width,
chosen number of columns and character type and size \(en as to have groff_ms break the lines and
`squeeze' the table between text margins.
.
.
.SH COMPATIBILITY
.PP
This program has been tested to run on both GNU/Linux (Ubuntu 22.04, Tiny Core Linux 10, Alpine Linux 3.12
within iSH app on iOS smartphone), as well as on MacOS X.
.
.
.
.SH BUGS
No known bugs.
.
.
.SH AUTHOR
Written by Rob Toscani (rob_toscani@yahoo.com)
.
.
.
.SH REPORTING BUGS
Please report any bugs to the author by e-mail or via https://github.com/jazzfan2/text2troff/issues


text2troff_table.sh
#!/bin/bash
# Name  : text2troff_table.sh
# Author: R.J.Toscani
# Date  : 26-10-2023
# Description: 'text2troff_table.sh' is a bash script that uses awk filtering to convert 
# tabulated lines in a plain-text-file into TROFF formatted tables by inserting table requests
# and ms macros. All lines not interpreted as table lines are passed on unchanged. The result
# is sent to standard output, and can be processed further with available 'troff' tools with
# 'ms' macro package, preferrably GNU groff_ms(7), to produce typeset PostScript-, PDF-, 
# HTML- or terminal ('nroff') output. As an alternative to processing a text-file, 
# text2troff_table.sh can also read (text) input from a pipe.
#
# (Comment texts are still partly in Dutch - will be translated into English in due course!)
#
######################################################################################
#
# Copyright (C) 2024 Rob Toscani <rob_toscani@yahoo.com>

# text2troff_table,sh is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# text2troff_table.sh is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

#######################################################################################
#
set -o noclobber    # Prevent files to be accidentally overwritten by redirection of standard output


helptext()
# Text printed if -h option (help) or a non-existent option has been given:
{
    while read "line"; do
        echo "$line" >&2         # print to standard error (stderr)
    done << EOF

Usage:
text2troff_table.sh [-ehs] TEXTFILE

   -e     Stretch tables to full text column width ('expand')
   -s     Force tables within text column width("squeeze") if too wide
   -h     Help (this output)

EOF
}


noreturn()
{
	cat "$1" | tr -d '\r' 
}


transcode()
# Stuurt de inhoud van een ISO-8859 tekstfile naar stdout in UTF-8 formaat (om het "grepbaar" te maken):
# https://www.tecmint.com/convert-files-to-utf-8-encoding-in-linux/
# De functie verwerkt een tekstfile die wordt opgegeven als argument, maar kan ook standard output
# verwerken die wordt aangeboden via een pipe.
{
    text="./text$RANDOM"
    cat "$1" >| $text
    if ( command -v file 1>/dev/null && command -v iconv 1>/dev/null ); then 
        file "$text" |
        if grep -q "ISO-8859"; then
#           iconv -f ISO-8859-1 -t UTF-8//TRANSLIT $text
            iconv -f ISO-8859-1 -t UTF-8 $text
        else
            cat $text
        fi
    else
        cat $text
    fi
    \rm $text
}


cleanup()
# Remove temporary files:
{
    [ -f "$textfile" ] && \rm $textfile
    [ -f "$columncount_file" ] && \rm $columncount_file
    exit
}


trap "cleanup; exit" SIGINT

if [[ -d /tmp/ramdisk/ ]]; then
    tmpfiledir="/tmp/ramdisk"
elif [[ -d /dev/shm/ ]]; then
    tmpfiledir="/dev/shm"
else
    tmpfiledir="."
fi

textfile="$tmpfiledir/textfile$RANDOM"
columncount_file="$tmpfiledir/columncount$RANDOM"
textwidth=78      # Iets gesmokkeld t.o.v. werkelijke afstand linker tot rechtermarge in troff bij 10 pts = 74
tabletype="center"
squeeze=0

while getopts "ehs" OPTION; do
    case $OPTION in
        e) tabletype="expand"
           ;;
        h) helptext
           exit 0
           ;;
        s) squeeze=1
           ;;
        *) helptext
           exit 1
           ;;
    esac
done
shift $((OPTIND - 1))

[[ $# > 1 ]] && textwidth=$2


noreturn "$1" |	transcode - |

# Eerst wordt van alle niet-display regels elke opeenvolging van meer dan 1 tab beperkt tot 1 enkele tab.
# Ook worden eventueel nog aanwezige begin- en eindtabs verwijderd.
# Onderstaande nieuwe uitgangspunten liggen hieraan ten grondslag:
# Elke groep van opeenvolgende tabs in uitgelijnde tabellen betreft per definitie *ÉÉN (1)* kolom, 
# en moet dus eerst expliciet worden genormaliseerd naar 1 tab voor verdere verwerking.
# Dus in de hiervoor aangeleverde tabel gelden twee opeenvolgende tabs *PER DEFINITIE NIET MEER*
# als begrenzing van een daartussen liggende lege kolom !!
# Om een lege kolom te representeren moet als conventie vanaf nu een spatie (of een ander vrij te kiezen 
# karakter voor 'leeg' b.v. '-') worden geplaatst (in de laatste kolom volstaat een spatie in minimaal 1 regel).
# D.i. vóór en/of na altijd van elkaar *GESCHEIDEN* tabs, met inbegrip van de eerste en laatste kolom.
# Begin- en eind-tabs vallen daarmee buiten de definitie van (extra) cel/kolom, zijn daarmee ongeldig en
# worden door het programma verwijderd.
#
# Suggestie voor verbetering:
# Idealiter zou dit alleen moeten gebeuren bij echte tabelregels, niet bij overige regels.
# Dit om de overige tekst ongemoeid te laten.
#
awk 'BEGIN {display_started = 0}
/^\.DS/  { display_started = 1 }
/^\.DE$/ { display_started = 0 }
{
    if (! display_started){
        gsub(/\t+/, "\t")
        gsub(/^\t/, "")
        gsub(/\t$/, "")
    }
    print
}' |


# Vervolgens worden de eventuele .br regels tussen opeenvolgende tabelregels verwijderd (als deze al 
# in een TROFF tekst aanwezig zijn; tabel = twee opeenvolgende regels met minimaal twee *GESCHEIDEN* tabs):
#
awk 'BEGIN {FS = "\t"; prev = ""; pprev = ""}

{
    if ( ! ($0 ~ /\t[^\t]+\t/ && prev ~ /^\.br$/ && pprev ~ /\t[^\t]+\t/)) # "NAND"-conditie!
        print prev
    pprev = prev
    prev = $0
}

END {print prev}' |


# Van elke tabel wordt het grootste aantal cellen bepaald dat in regels van die tabel voorkomt.
# De gevonden waarde zal worden ingesteld als het aantal kolommen voor de tabel in kwestie.
#
awk 'BEGIN {FS = "\t"; tbl = 1; prev_line = ""; display_started = 0; rowcount = 0}
/^\.DS/  { display_started = 1 }
/^\.DE$/ { display_started = 0 }

$0 ~ /\t[^\t]+\t/ && ! display_started \
{
    if (prev_line !~ /\t[^\t]+\t/){       # Eerste regel van potentiële nieuwe tabel
        if(rowcount > 1)                  # Als de vorige tabel rowcount > 2 regels  had, dus een *echte* tabel was ...
            ++tbl                         # .. dan krijgt de nieuwe (potentiële) tabel een *HOGER* nummer (tbl)
        else                              # Anders was de vorige tabel *onecht* en wordt het nummer *HERGEBRUIKT*
            columncount[tbl] = 0          # ...en wordt van de "onechte" tabel de eerder gevulde arraywaarde weer genuld
        rowcount = 1                      # Elke potentiële tabel begint met 1 regel
    }
    else
        ++rowcount                        # Hier heeft de tabel al minimaal 2 regels
    if (NF > columncount[tbl])
        columncount[tbl] = NF             # Max. aantal velden per regel per tabel => dus het aantal tabelkolommen
}

{
    print
    prev_line = $0
}

# Wegschrijven van de 2D-array "columncount" naar een gelijknamige file, die aan het begin van het 
# volgende awk-blok weer zal worden ingelezen naar een vergelijkbare 2D-array:
#
END \
{
    for (tbl = 1; tbl <= length(columncount); tbl++)        # Denk erom: *EXPLICIET IN VOLGORDE* langs de indices in awk!
        print columncount[tbl] >> "'"$columncount_file"'"

}' >| "$textfile"


[[ ! -f "$columncount_file" ]] && cat "$textfile" && cleanup


# Vervolgens worden alle valide tabelregels voorzien van de benodigde macro's en formateringsregels
# om in opmaak te worden weergegeven indien via een pipe aangeboden aan (b.v.) trofms() of nrofms() of een 
# andere TROFF functie met "tbl" optie:
#
awk -v tabletype="$tabletype" -v squeeze="$squeeze" -v "textwidth=$textwidth" '\

    function max(a, b){
        if (a >= b)
            return a
        else
            return b
    }

BEGIN {
    tbl = 0
    while (getline < "'"$columncount_file"'"){     # Max. aantal velden per regel per tabel => dus het aantal tabelkolommen
        ++tbl
        columncount[tbl] = $0
    }
    close("'"$columncount_file"'")

    FS = "\t"
    tbl = 0
    prev_line = ""
    prev_row = ""
    format_on = 0
    table_started = 0
    display_started = 0
    print ".nr table_nr 0"                          # Nieuw number register "table_nr" wordt geïnitialiseerd op 0
    print ".nr ps_decr \\n[.s]*2/10"                # Nieuw number register "ps_decr" wordt op 20% van de point size gesteld
}


/^\.DS/  { display_started = 1 }
/^\.DE$/ { display_started = 0 }

/\t[^\t]+\t/ && ! display_started \
{
    if (prev_line !~ /\t[^\t]*\t/){                 # De huidige regel is de eerste van een *potentiële* tabel
        type = tabletype                            # Maak type gelijk aan de ingestelde tabletype 
        format_string2 = ""
        for (i = 1; i <= columncount[tbl + 1]; i++) # Kijk alvast "vooruit" om de bijbehorende column-count op te halen
            format_string2 = "l "format_string2     # Leg deze vast in een string, die echter nog niet wordt uitgeprint
        format_on = 1                               # 1e regel is nu gesignaleerd, maar "tbl" kan pas omhoog bij 2e regel
        if (squeeze){
            chunk = int(max(textwidth/columncount[tbl+1]-4, 1)) # Kijk alvast "vooruit" om de bijbehorende "chunk" te bepalen
            regex = "("
            for (i = 1; i <= chunk; i++)            # Matchend aantal karakters, niet \:[( om escape-collisions te beperken
                regex = regex"[^:*\\\\\\[\\( ]"     # <============ OS X, hierin werken {} quantifiers niet ===============
            regex = regex")"
        }

    }
    else{
        if (format_on){                             # Dit is dus de 2e tabelregel, teken dat de tabel nu pas "echt" is
            ++tbl                                   # Hier kan dus pas het tabel-nummer omhoog
            print ".LP"                             # Bedoeld tegen table-orphans. Werkt niet want begin tabel = eind alinea
            print ".nr table_nr +1"                 # Hier laten we de waarde van het troff-register "table_nr" verhogen
            print ".ps -\\n[ps_decr]"               # en de point size verminderen met register "ps_decr" = 20% * point size
            print ".TS H"
            maxchunk = max(textwidth/8, 6)
            if (squeeze && chunk < maxchunk)        # Bij "squeeze" en onder een empirisch bepaalde maximale chunkwaarde ...
                type = "expand"                     # ... wijzig dan type (alsnog) van "center" naar "expand"
            print type", allbox;"
            format_string1 = "c"
            for (i = 2; i <= columncount[tbl]; ++i)
                format_string1 = format_string1 " s"
            print format_string1                    # Nu kunnen de format strings (incl de eerder voorbereide) worden geprint
            print format_string2 "."
            print ".ft 3"                           # Bolt font
            print "Tabel \\n[table_nr]"             # Tekst "Tabel" met huidige waarde van register "table_nr"
            print ".P"                              # Previous font
            print "_"
            print ".TH"
            format_on = 0                           # "format_on = 1" was een tussenstadium, dat is nu geëindigd
            table_started = 1                       # Tabel is gestart
        }
    }
    row = ""
    i = 1
    while (length($i)){
        text = $i
        if (squeeze)
            gsub(regex, "&\\:", text)               # Voeg een \: break sequence in na elke "chunk" aan karakters.
        row = row"T{\n"text"\nT}\t"
        i++
    }
}

prev_line !~ /\t[^\t]+\t/ && table_started \
{
    print ".TE"
    print ".LP"
    table_started = 0
}

{
    if (table_started)
        print prev_row
    else
        print prev_line
    prev_line = $0
    prev_row = row
}

END {
    if (prev_line !~ /\t[^\t]*\t/ && table_started){
        print ".TE"
        print ".LP"
        print prev_line
    }
    else if (prev_line ~ /\t[^\t]*\t/ && table_started){
        print row
        print ".TE"
    }
    else
        print prev_line
 }' "$textfile"

cleanup


# ==================== RJT SOFTWARE LABS - FOR MIRACLES IN SYSTEM DEVELOPMENT ====================


